vllm/entrypoints/openai/api_server.py

```diff
@@ -472,36 +502,46 @@ def create_stream_response_json(
     async def completion_stream_generator() -> AsyncGenerator[str, None]:
         previous_texts = [""] * request.n
         previous_num_tokens = [0] * request.n
+        has_echoed = [False] * request.n
         async for res in result_generator:
             res: RequestOutput
             for output in res.outputs:
                 i = output.index
                 delta_text = output.text[len(previous_texts[i]):]
+                token_ids = output.token_ids[previous_num_tokens[i]:]
+                top_logprobs = output.logprobs[previous_num_tokens[i]:]
+                offsets = len(previous_texts[i])
+                if request.echo and not has_echoed[i]:
+                    if not echo_self:
+                        delta_text = res.prompt + delta_text
+                        token_ids = res.prompt_token_ids + token_ids
+                        top_logprobs = res.prompt_logprobs + top_logprobs
+                    else:
+                        delta_text = res.prompt
+                        token_ids = res.prompt_token_ids
+                        top_logprobs = res.prompt_logprobs
+                    has_echoed[i] = True
                 if request.logprobs is not None:
                     logprobs = create_logprobs(
-                        output.token_ids[previous_num_tokens[i]:],
-                        output.logprobs[previous_num_tokens[i]:],
-                        len(previous_texts[i]))
+                        token_ids=token_ids,
+                        top_logprobs=top_logprobs,
+                        num_output_top_logprobs=request.logprobs,
+                        initial_text_offset=offsets,
+                    )
                 else:
                     logprobs = None
                 previous_texts[i] = output.text
                 previous_num_tokens[i] = len(output.token_ids)
+                finish_reason = output.finish_reason
+                if output.finish_reason is None and echo_self and has_echoed[i]:
+                    finish_reason = "length"
                 response_json = create_stream_response_json(
                     index=i,
                     text=delta_text,
                     logprobs=logprobs,
+                    finish_reason=finish_reason,
                 )
                 yield f"data: {response_json}\n\n"
-                if output.finish_reason is not None:
-                    logprobs = (LogProbs()
-                                if request.logprobs is not None else None)
-                    response_json = create_stream_response_json(
-                        index=i,
-                        text="",
-                        logprobs=logprobs,
-                        finish_reason=output.finish_reason,
-                    )
-                    yield f"data: {response_json}\n\n"
```

Comment:
This is to mimic OpenAI server's behavior. Please add it back. When query the server with:
```python
completion = openai.Completion.create(
    model="text-davinci-003", prompt="Say this is a test", n=1, echo=True, best_of=1,
    stream=True, logprobs=2, max_tokens=5)
for c in completion:
    print(c)
```
The reply is:
```
{
  "choices": [
    {
      "finish_reason": null,
      "index": 0,
      "logprobs": {
        "text_offset": [
          0,
          3,
          8,
          11,
          13,
          18
        ],
        "token_logprobs": [
          null,
          -5.8172827,
          -2.688938,
          -1.3986964,
          -3.46766,
          -0.05558832
        ],
        "tokens": [
          "Say",
          " this",
          " is",
          " a",
          " test",
          "\n"
        ],
        "top_logprobs": [
          null,
          {
            "\n": -2.923694,
            "(\"": -3.1542532
          },
          {
            "\n": -0.79770833,
            " is": -2.688938
          },
          {
            " a": -1.3986964,
            " the": -1.8430934
          },
          {
            " test": -3.46766,
            " very": -3.8154793
          },
          {
            "\n": -0.05558832,
            ".": -4.5110774
          }
        ]
      },
      "text": "Say this is a test\n"
    }
  ],
  "created": 1698764284,
  "id": "cmpl-8FkTsfS1sPFzyue9mEKjl87ZtPmt7",
  "model": "text-davinci-003",
  "object": "text_completion",
  "warning": "This model version is deprecated. Migrate before January 4, 2024 to avoid disruption of service. Learn more https://platform.openai.com/docs/deprecations"
}
{
  "choices": [
    {
      "finish_reason": null,
      "index": 0,
      "logprobs": {
        "text_offset": [
          19
        ],
        "token_logprobs": [
          -0.01684946
        ],
        "tokens": [
          "\n"
        ],
        "top_logprobs": [
          {
            "\n": -0.01684946,
            "Say": -4.6857796
          }
        ]
      },
      "text": "\n"
    }
  ],
  "created": 1698764284,
  "id": "cmpl-8FkTsfS1sPFzyue9mEKjl87ZtPmt7",
  "model": "text-davinci-003",
  "object": "text_completion",
  "warning": "This model version is deprecated. Migrate before January 4, 2024 to avoid disruption of service. Learn more https://platform.openai.com/docs/deprecations"
}
{
  "choices": [
    {
      "finish_reason": null,
      "index": 0,
      "logprobs": {
        "text_offset": [
          20
        ],
        "token_logprobs": [
          -0.54426545
        ],
        "tokens": [
          "This"
        ],
        "top_logprobs": [
          {
            "This": -0.54426545,
            "Yes": -0.95614594
          }
        ]
      },
      "text": "This"
    }
  ],
  "created": 1698764284,
  "id": "cmpl-8FkTsfS1sPFzyue9mEKjl87ZtPmt7",
  "model": "text-davinci-003",
  "object": "text_completion",
  "warning": "This model version is deprecated. Migrate before January 4, 2024 to avoid disruption of service. Learn more https://platform.openai.com/docs/deprecations"
}
{
  "choices": [
    {
      "finish_reason": null,
      "index": 0,
      "logprobs": {
        "text_offset": [
          24
        ],
        "token_logprobs": [
          -0.047382396
        ],
        "tokens": [
          " is"
        ],
        "top_logprobs": [
          {
            " indeed": -3.6144009,
            " is": -0.047382396
          }
        ]
      },
      "text": " is"
    }
  ],
  "created": 1698764284,
  "id": "cmpl-8FkTsfS1sPFzyue9mEKjl87ZtPmt7",
  "model": "text-davinci-003",
  "object": "text_completion",
  "warning": "This model version is deprecated. Migrate before January 4, 2024 to avoid disruption of service. Learn more https://platform.openai.com/docs/deprecations"
}
{
  "choices": [
    {
      "finish_reason": "length",
      "index": 0,
      "logprobs": {
        "text_offset": [
          27
        ],
        "token_logprobs": [
          -0.028564425
        ],
        "tokens": [
          " indeed"
        ],
        "top_logprobs": [
          {
            " a": -3.7224274,
            " indeed": -0.028564425
          }
        ]
      },
      "text": " indeed"
    }
  ],
  "created": 1698764284,
  "id": "cmpl-8FkTsfS1sPFzyue9mEKjl87ZtPmt7",
  "model": "text-davinci-003",
  "object": "text_completion",
  "warning": "This model version is deprecated. Migrate before January 4, 2024 to avoid disruption of service. Learn more https://platform.openai.com/docs/deprecations"
}
{
  "choices": [
    {
      "finish_reason": "length",
      "index": 0,
      "logprobs": {
        "text_offset": [],
        "token_logprobs": [],
        "tokens": [],
        "top_logprobs": []
      },
      "text": ""
    }
  ],
  "created": 1698764284,
  "id": "cmpl-8FkTsfS1sPFzyue9mEKjl87ZtPmt7",
  "model": "text-davinci-003",
  "object": "text_completion",
  "warning": "This model version is deprecated. Migrate before January 4, 2024 to avoid disruption of service. Learn more https://platform.openai.com/docs/deprecations"
}

```
Note in the very last json reply it only includes the finish reason.

Reply:
Interesting... It looks like I tried other OpenAI models, and they behaved in a different way. Adding them back.